p2p/src/net/mock/transport/adapted.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod identity;
+pub mod noise;
+pub mod traits;
+
+use std::{sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use tokio::{sync::mpsc::UnboundedReceiver, time::timeout};
+
+use crate::{error::P2pError, net::mock::peer::Role, Result};
+
+use self::traits::StreamAdapter;
+
+use super::{TransportListener, TransportSocket};
+
+// How much time is allowed to spend setting up (optionally) encrypted stream.
+const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// Transport layer that wraps a lower-level transport layer (can be seen like an onion with multiple layer)
+/// Simplest version of this can be seen as a tcp transport layer, with an Identity stream_adapter. That would
+/// be equivalent to the tcp transport layer with nothing done to it.
+/// More layers can be added on top of this, with this struct, where we add encryption on top.
+#[derive(Debug)]
+pub struct WrappedTransportSocket<S, T> {
+    stream_adapter: Arc<S>,
+    base_transport: T,
```

Comment:
One conclusion from our call is that having this owned by WrappedTransportSocket will result in the initialization of the base_transport being monolithic, which will make it impossible for keys to be imported from disk. This is a problem that already extends to the current implementation in libp2p, where the keys are also automatically generated. The keys being automatically (newly) generated on every connection is the only reason this is possible in its current form. Generally speaking, every stream layer should hand its prepared state to its parent in an owning fashion, and then the highest layer should go into the transport layer. This makes it possible to initialize keys separately for every layer in different ways, instead of in the current monolithic fashion.

This doesn't have to be fixed in this PR since this will touch the libp2p implementation, but this has to be fixed next.

Reply:
OK, thank you for the summary. Let's do that in later PR.